app/Main.hs

```diff
@@ -23,61 +33,82 @@ data Command = MoveRight
              deriving (Show, Eq)
 
 tokenize :: String -> [Command]
-tokenize [] = []
-tokenize (x:xs)
-  | x == '>' = MoveRight : tokenize xs
-  | x == '<' = MoveLeft : tokenize xs
-  | x == '+' = IncCell: tokenize xs
-  | x == '-' = DecCell : tokenize xs
-  | x == '.' = PrintCell : tokenize xs
-  | x == ',' = InputCell: tokenize xs
-  | x == '[' = OpenBracket : tokenize xs
-  | x == ']' = CloseBracket: tokenize xs
-  | otherwise = tokenize xs
+tokenize = catMaybes . fmap charToCommand
+  where
+    charToCommand c =
+      case c of
+        '>' -> Just MoveRight
+        '<' -> Just MoveLeft
+        '+' -> Just IncCell
+        '-' -> Just DecCell
+        '.' -> Just PrintCell
+        ',' -> Just InputCell
+        '[' -> Just OpenBracket
+        ']' -> Just CloseBracket
+        _ -> Nothing
 
-
--- Sets a specific cell to a value
 setCell :: (Int -> Int) -> CellState -> Int -> CellState
-setCell f cs i = take i cs ++ [if j < 0 then 255 else j] ++ drop (i + 1) cs
```

Comment:
`take` scans the list till you get to `i`. `drop` scans the list till you get to `i+1`. `(!!)` scans the list till it gets to `i`. You actually end up scanning over the same elements three times when you use `take`, `(!!)` and `drop` like this.